aten/src/ATen/native/ForeachUtils.h

```diff
@@ -88,61 +88,46 @@ bool has_same_attributes(Device expected_device, TensorList tensors) {
   return true;
 }
 
-bool will_promote_tensor(const Tensor& tensor, Scalar scalar) {
-  auto result_dtype = at::result_type(tensor, scalar);
-  return result_dtype != tensor.scalar_type();
-}
-
-bool can_use_fast_route(TensorList tensors) {
-#ifdef __HIP_PLATFORM_HCC__
-  return false;
-#else
-  auto expected_device = tensors[0].device();
-  for (auto t : tensors) {
-    if (!has_same_attributes(expected_device, {t})) {
-      return false;
+bool will_promote_tensor(const Tensor& tensor, Scalar scalar, bool promote_integer_inputs_to_float = false) {
+  // In case of division, integer inputs will result in float
```

Comment:
a fair number of unary operations now supports integer to float promotion, grep for `promotes_integers_to_float` in common_methods_invocations.py, which means that you'd need to disable them for a fast path. In general, keeping track of this is pretty hard, so to make it more automatic perhaps testing `for_each` ops should also become OpInfo based? At least supported or unsupported dtypes, we can discuss in more details.

Reply:
Added a few PRs to this stack that will move tests to OpInfo